lib/assert.js

```diff
@@ -475,24 +473,59 @@ function partiallyCompareArrayBuffersOrViews(actual, expected) {
 }
 
 function partiallyCompareSets(actual, expected, comparedObjects) {
-  if (SetPrototypeGetSize(expected) > SetPrototypeGetSize(actual)) {
-    return false; // `expected` can't be a subset if it has more elements
+  if (actual === expected) return true;
+
+  const expectedSize = SetPrototypeGetSize(expected);
+  if (expectedSize === 0) return true;
+
+  const actualSize = SetPrototypeGetSize(actual);
+  if (expectedSize > actualSize) {
+    return false;
   }
 
   if (isDeepEqual === undefined) lazyLoadComparison();
 
-  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
+  // Separate expected items into direct matches and those needing deep comparison
+  const expectedItemsToCompare = []; // Items that need deep comparison
   const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
-  const usedIndices = new SafeSet();
 
-  expectedIteration: for (const expectedItem of expectedIterator) {
-    for (let actualIdx = 0; actualIdx < actualArray.length; actualIdx++) {
-      if (!usedIndices.has(actualIdx) && isDeepStrictEqual(actualArray[actualIdx], expectedItem)) {
-        usedIndices.add(actualIdx);
-        continue expectedIteration;
+  for (const expectedItem of expectedIterator) {
+    if (actual.has(expectedItem)) {
+      continue;
+    } else if (isPrimitive(expectedItem)) {
+      return false;
+    } else {
+      expectedItemsToCompare.push(expectedItem);
+    }
+  }
+
+  if (expectedItemsToCompare.length === 0) {
+    return true;
+  }
+
+  // Only collect non-primitive items from actual for deep comparison
+  const actualSet = new SafeSet();
+  for (const item of FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual)) {
+    if (!isPrimitive(item)) {
+      actualSet.add(item);
+    }
+  }
+
+  // Now do deep comparison for the remaining expected items
+  for (const expectedItem of expectedItemsToCompare) {
+    let foundMatch = false;
+
+    for (const actualItem of actualSet) {
+      if (isDeepStrictEqual(actualItem, expectedItem)) {
+        actualSet.delete(actualItem);
+        foundMatch = true;
+        break;
       }
     }
-    return false;
+
+    if (!foundMatch) {
+      return false;
+    }
   }
```

Comment:
These can be combined into a single loop where we do not need to push any entries into the actualSet by not using an array initially but a set as it was originally.
Now we just have a single loop over the actualSet and if it's not a primitive, that's compared with entries in the expectedObjectSet.
This removes the need for the inner for loop here completely. So we safe up to (n - 1) * m iterations.
Please check the code in the original comparison method. It is used there. Maybe also compare performance for both implementations to verify the difference.

Reply:
🤦 you are right, I am doing too many things at the same time.